File: idea/src/org/jetbrains/jet/plugin/debugger/JetPositionManager.java
/*
 * Copyright 2010-2013 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.jet.plugin.debugger;

import com.intellij.debugger.NoDataException;
import com.intellij.debugger.PositionManager;
import com.intellij.debugger.SourcePosition;
import com.intellij.debugger.engine.DebugProcess;
import com.intellij.debugger.requests.ClassPrepareRequestor;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.util.Ref;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.util.*;
import com.sun.jdi.AbsentInformationException;
import com.sun.jdi.Location;
import com.sun.jdi.ReferenceType;
import com.sun.jdi.request.ClassPrepareRequest;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.TestOnly;
import org.jetbrains.org.objectweb.asm.Type;
import org.jetbrains.jet.analyzer.AnalyzeExhaust;
import org.jetbrains.jet.codegen.ClassBuilderMode;
import org.jetbrains.jet.codegen.PackageCodegen;
import org.jetbrains.jet.codegen.binding.CodegenBinding;
import org.jetbrains.jet.codegen.state.JetTypeMapper;
import org.jetbrains.jet.lang.descriptors.ClassDescriptor;
import org.jetbrains.jet.lang.descriptors.ValueParameterDescriptor;
import org.jetbrains.jet.lang.psi.*;
import org.jetbrains.jet.lang.resolve.BindingContext;
import org.jetbrains.jet.lang.resolve.DelegatingBindingTrace;
import org.jetbrains.jet.lang.resolve.calls.model.ResolvedCall;
import org.jetbrains.jet.lang.resolve.calls.model.ResolvedValueArgument;
import org.jetbrains.jet.lang.resolve.extension.InlineAnalyzerExtension;
import org.jetbrains.jet.lang.resolve.java.JetFilesProvider;
import org.jetbrains.jet.lang.resolve.java.JvmClassName;
import org.jetbrains.jet.lang.resolve.name.FqName;
import org.jetbrains.jet.lang.types.lang.InlineStrategy;
import org.jetbrains.jet.lang.types.lang.InlineUtil;
import org.jetbrains.jet.plugin.project.AnalyzerFacadeWithCache;
import org.jetbrains.jet.plugin.util.DebuggerUtils;

import java.util.*;

import static org.jetbrains.jet.codegen.binding.CodegenBinding.asmTypeForAnonymousClass;

public class JetPositionManager implements PositionManager {
    private final DebugProcess myDebugProcess;
    private final WeakHashMap<FqName, CachedValue<JetTypeMapper>> myTypeMappers = new WeakHashMap<FqName, CachedValue<JetTypeMapper>>();

    public JetPositionManager(DebugProcess debugProcess) {
        myDebugProcess = debugProcess;
    }

    @Override
    @Nullable
    public SourcePosition getSourcePosition(@Nullable Location location) throws NoDataException {
        if (location == null) {
            throw new NoDataException();
        }
        PsiFile psiFile = getPsiFileByLocation(location);
        if (psiFile == null) {
            throw new NoDataException();
        }

        int lineNumber;
        try {
            lineNumber = location.lineNumber() - 1;
        }
        catch (InternalError e) {
            lineNumber = -1;
        }

        if (lineNumber >= 0) {
            return SourcePosition.createFromLine(psiFile, lineNumber);
        }

        throw new NoDataException();
    }

    @Nullable
    private PsiFile getPsiFileByLocation(@NotNull Location location) {
        String sourceName;
        try {
            sourceName = location.sourceName();
        }
        catch (AbsentInformationException e) {
            return null;
        }

        // JDI names are of form "package.Class$InnerClass"
        String referenceFqName = location.declaringType().name();
        String referenceInternalName = referenceFqName.replace('.', '/');
        JvmClassName className = JvmClassName.byInternalName(referenceInternalName);

        return DebuggerUtils.findSourceFileForClass(GlobalSearchScope.allScope(myDebugProcess.getProject()), className, sourceName);
    }

    @NotNull
    @Override
    public List<ReferenceType> getAllClasses(SourcePosition sourcePosition) throws NoDataException {
        if (!(sourcePosition.getFile() instanceof JetFile)) {
            throw new NoDataException();
        }
        String name = classNameForPosition(sourcePosition);
        List<ReferenceType> result = new ArrayList<ReferenceType>();
        if (name != null) {
            result.addAll(myDebugProcess.getVirtualMachineProxy().classesByName(name));
        }
        return result;
    }

    @Nullable
    private String classNameForPosition(final SourcePosition sourcePosition) {
        final Ref<String> result = Ref.create();

        ApplicationManager.getApplication().runReadAction(new Runnable() {
            @Override
            @SuppressWarnings("unchecked")
            public void run() {
                JetFile file = (JetFile) sourcePosition.getFile();
                JetTypeMapper typeMapper = prepareTypeMapper(file);

                result.set(getClassNameForElement(sourcePosition.getElementAt(), typeMapper, file));
            }

        });

        return result.get();
    }

    @SuppressWarnings("unchecked")
    private static String getClassNameForElement(@Nullable PsiElement notPositionedElement, @NotNull JetTypeMapper typeMapper, @NotNull JetFile file) {
        PsiElement element = PsiTreeUtil.getParentOfType(notPositionedElement, JetClassOrObject.class, JetFunctionLiteral.class, JetNamedFunction.class);

        if (element instanceof JetClassOrObject) {
            return getJvmInternalNameForImpl(typeMapper, (JetClassOrObject) element);
        }
        else if (element instanceof JetFunctionLiteral) {
            if (isInlinedLambda((JetFunctionLiteral) element, typeMapper.getBindingContext())) {
                return getClassNameForElement(element.getParent(), typeMapper, file);
            } else {
                Type asmType = asmTypeForAnonymousClass(typeMapper.getBindingContext(), ((JetFunctionLiteral) element));
                return asmType.getInternalName();
            }
        }
        else if (element instanceof JetNamedFunction) {
            PsiElement parent = PsiTreeUtil.getParentOfType(element, JetClassOrObject.class, JetFunctionLiteralExpression.class, JetNamedFunction.class);
            if (parent instanceof JetClassOrObject) {
                return getJvmInternalNameForImpl(typeMapper, (JetClassOrObject) parent);
            }
            else if (parent instanceof JetFunctionLiteralExpression || parent instanceof JetNamedFunction) {
                Type asmType = asmTypeForAnonymousClass(typeMapper.getBindingContext(), (JetElement) element);
                return asmType.getInternalName();
            }
        }

        return PackageCodegen.getPackagePartInternalName(file);
    }


    @Nullable
    private static String getJvmInternalNameForImpl(JetTypeMapper typeMapper, JetClassOrObject jetClass) {
        ClassDescriptor classDescriptor = typeMapper.getBindingContext().get(BindingContext.CLASS, jetClass);
        if (classDescriptor == null) {
            return null;
        }

        if (jetClass instanceof JetClass && ((JetClass) jetClass).isTrait()) {
            return typeMapper.mapTraitImpl(classDescriptor).getInternalName();
        }

        return typeMapper.mapClass(classDescriptor).getInternalName();
    }

    private JetTypeMapper prepareTypeMapper(final JetFile file) {
        FqName fqName = JetPsiUtil.getFQName(file);
        CachedValue<JetTypeMapper> value = myTypeMappers.get(fqName);
        if(value == null) {
            value = CachedValuesManager.getManager(file.getProject()).createCachedValue(new CachedValueProvider<JetTypeMapper>() {
                @Override
                public Result<JetTypeMapper> compute() {
                    AnalyzeExhaust analyzeExhaust = AnalyzerFacadeWithCache.analyzeFileWithCache(file);
                    analyzeExhaust.throwIfError();

                    Collection<JetFile> packageFiles = JetFilesProvider.getInstance(file.getProject()).allPackageFiles().fun(file);

                    DelegatingBindingTrace bindingTrace = new DelegatingBindingTrace(analyzeExhaust.getBindingContext(), "trace created in JetPositionManager");
                    JetTypeMapper typeMapper = new JetTypeMapper(bindingTrace, ClassBuilderMode.FULL);
                    //noinspection unchecked
                    CodegenBinding.initTrace(bindingTrace, packageFiles);
                    return new Result<JetTypeMapper>(typeMapper, PsiModificationTracker.MODIFICATION_COUNT);
                }
            }, false);
            myTypeMappers.put(fqName, value);
        }

        return value.getValue();
    }

    @NotNull
    @Override
    public List<Location> locationsOfLine(ReferenceType type, SourcePosition position) throws NoDataException {
        if (!(position.getFile() instanceof JetFile)) {
            throw new NoDataException();
        }
        try {
            int line = position.getLine() + 1;
            List<Location> locations = myDebugProcess.getVirtualMachineProxy().versionHigher("1.4")
                                       ? type.locationsOfLine(DebugProcess.JAVA_STRATUM, null, line)
                                       : type.locationsOfLine(line);
            if (locations == null || locations.isEmpty()) throw new NoDataException();
            return locations;
        }
        catch (AbsentInformationException e) {
            throw new NoDataException();
        }
    }

    @Override
    public ClassPrepareRequest createPrepareRequest(ClassPrepareRequestor classPrepareRequestor,
                                                    SourcePosition sourcePosition) throws NoDataException {
        if (!(sourcePosition.getFile() instanceof JetFile)) {
            throw new NoDataException();
        }
        String className = classNameForPosition(sourcePosition);
        if (className == null) {
            return null;
        }
        return myDebugProcess.getRequestsManager().createClassPrepareRequest(classPrepareRequestor, className.replace('/', '.'));
    }

    @TestOnly
    public void addTypeMapper(JetFile file, final JetTypeMapper typeMapper) {
        FqName fqName = JetPsiUtil.getFQName(file);
        CachedValue<JetTypeMapper> value = CachedValuesManager.getManager(file.getProject()).createCachedValue(new CachedValueProvider<JetTypeMapper>() {
            @Override
            public Result<JetTypeMapper> compute() {
                return new Result<JetTypeMapper>(typeMapper, PsiModificationTracker.MODIFICATION_COUNT);
            }
        }, false);
        myTypeMappers.put(fqName, value);
    }


    public static boolean isInlinedLambda(@NotNull JetFunctionLiteral functionLiteral, @NotNull BindingContext context) {
        PsiElement functionLiteralExpression = functionLiteral.getParent();
        if (functionLiteralExpression == null) return false;

        PsiElement parent = functionLiteralExpression.getParent();

        PsiElement valueArgument = functionLiteralExpression;
        while (parent instanceof JetParenthesizedExpression ||
               parent instanceof JetBinaryExpressionWithTypeRHS ||
               parent instanceof JetPrefixExpression && JetPsiUtil.isLabeledExpression((JetPrefixExpression) parent)) {
            valueArgument = parent;
            parent = parent.getParent();
        }

        while (parent instanceof ValueArgument ||
               parent instanceof JetValueArgumentList) {
            parent = parent.getParent();
        }

        if ((parent == null || !(parent instanceof JetBinaryExpression)) && !(parent instanceof JetCallExpression)) return false;

        ResolvedCall<?> call = null;
        if (parent instanceof JetCallExpression) {
            call = context.get(BindingContext.RESOLVED_CALL, ((JetCallExpression) parent).getCalleeExpression());
        }
        if (parent instanceof JetBinaryExpression) {
            call = context.get(BindingContext.RESOLVED_CALL, ((JetBinaryExpression) parent).getOperationReference());
        }

        if (call == null) return false;

        InlineStrategy inlineType = InlineUtil.getInlineType(call.getResultingDescriptor().getAnnotations());
        if (!inlineType.isInline()) return false;

        for (Map.Entry<ValueParameterDescriptor, ResolvedValueArgument> entry : call.getValueArguments().entrySet()) {
            ValueParameterDescriptor valueParameterDescriptor = entry.getKey();
            ResolvedValueArgument resolvedValueArgument = entry.getValue();

            for (ValueArgument next : resolvedValueArgument.getArguments()) {
                JetExpression expression = next.getArgumentExpression();
                if (valueArgument == expression) {
                    return InlineAnalyzerExtension.checkInlinableParameter(
                            valueParameterDescriptor, expression,
                            call.getResultingDescriptor(), null
                    );
                }
            }
        }
        return false;
    }

}

File: idea/src/org/jetbrains/kotlin/idea/debugger/JetPositionManager.kt
/*
 * Copyright 2010-2015 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.kotlin.idea.debugger

import com.intellij.debugger.NoDataException
import com.intellij.debugger.PositionManager
import com.intellij.debugger.SourcePosition
import com.intellij.debugger.engine.DebugProcess
import com.intellij.debugger.requests.ClassPrepareRequestor
import com.intellij.openapi.project.DumbService
import com.intellij.openapi.roots.libraries.LibraryUtil
import com.intellij.openapi.util.Pair
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.util.*
import com.sun.jdi.AbsentInformationException
import com.sun.jdi.Location
import com.sun.jdi.ReferenceType
import com.sun.jdi.request.ClassPrepareRequest
import org.jetbrains.annotations.TestOnly
import org.jetbrains.kotlin.codegen.AsmUtil
import org.jetbrains.kotlin.codegen.ClassBuilderFactories
import org.jetbrains.kotlin.codegen.state.GenerationState
import org.jetbrains.kotlin.codegen.state.JetTypeMapper
import org.jetbrains.kotlin.descriptors.ClassDescriptor
import org.jetbrains.kotlin.descriptors.PropertyDescriptor
import org.jetbrains.kotlin.psi.*
import org.jetbrains.kotlin.resolve.BindingContext
import org.jetbrains.kotlin.resolve.calls.callUtil.*
import org.jetbrains.kotlin.resolve.extension.InlineAnalyzerExtension
import org.jetbrains.kotlin.resolve.jvm.JvmClassName
import org.jetbrains.kotlin.load.kotlin.PackagePartClassUtils
import org.jetbrains.kotlin.name.FqName
import org.jetbrains.kotlin.builtins.InlineUtil
import org.jetbrains.kotlin.idea.caches.resolve.*
import org.jetbrains.kotlin.idea.codeInsight.CodeInsightUtils
import org.jetbrains.kotlin.idea.util.DebuggerUtils

import java.util.*

import org.jetbrains.kotlin.codegen.binding.CodegenBinding.asmTypeForAnonymousClass
import org.jetbrains.kotlin.idea.stubindex.PackageIndexUtil
import org.jetbrains.kotlin.idea.util.application.runReadAction

public class JetPositionManager(private val myDebugProcess: DebugProcess) : PositionManager {
    private val myTypeMappers = WeakHashMap<Pair<FqName, IdeaModuleInfo>, CachedValue<JetTypeMapper>>()

    override fun getSourcePosition(location: Location?): SourcePosition? {
        if (location == null) {
            throw NoDataException()
        }
        val psiFile = getPsiFileByLocation(location)
        if (psiFile == null) {
            throw NoDataException()
        }

        val lineNumber = try {
            location.lineNumber() - 1
        }
        catch (e: InternalError) {
            -1
        }


        if (lineNumber >= 0) {
            val lambdaIfInside = getLambdaIfInside(location, psiFile as JetFile, lineNumber)
            if (lambdaIfInside != null) {
                return SourcePosition.createFromElement(lambdaIfInside.getBodyExpression().getStatements().get(0))
            }
            return SourcePosition.createFromLine(psiFile, lineNumber)
        }

        throw NoDataException()
    }

    private fun getLambdaIfInside(location: Location, file: JetFile, lineNumber: Int): JetFunctionLiteral? {
        val currentLocationFqName = location.declaringType().name()
        if (currentLocationFqName == null) return null

        val start = CodeInsightUtils.getStartLineOffset(file, lineNumber)
        val end = CodeInsightUtils.getEndLineOffset(file, lineNumber)
        if (start == null || end == null) return null

        val literals = CodeInsightUtils.findElementsOfClassInRange(file, start, end, javaClass<JetFunctionLiteral>())
        if (literals == null || literals.size() == 0) return null

        val isInLibrary = LibraryUtil.findLibraryEntry(file.getVirtualFile(), file.getProject()) != null
        val typeMapper = if (!isInLibrary)
            prepareTypeMapper(file)
        else
            createTypeMapperForLibraryFile(file.findElementAt(start), file)

        val currentLocationClassName = JvmClassName.byFqNameWithoutInnerClasses(FqName(currentLocationFqName)).getInternalName()
        for (literal in literals) {
            val functionLiteral = literal as JetFunctionLiteral
            if (isInlinedLambda(functionLiteral, typeMapper.getBindingContext())) {
                continue
            }

            val internalClassName = getClassNameForElement(literal.getFirstChild(), typeMapper, file, isInLibrary)
            if (internalClassName == currentLocationClassName) {
                return functionLiteral
            }
        }

        return null
    }

    private fun getPsiFileByLocation(location: Location): PsiFile? {
        val sourceName: String
        try {
            sourceName = location.sourceName()
        }
        catch (e: AbsentInformationException) {
            return null
        }


        // JDI names are of form "package.Class$InnerClass"
        val referenceFqName = location.declaringType().name()
        val referenceInternalName = referenceFqName.replace('.', '/')
        val className = JvmClassName.byInternalName(referenceInternalName)

        val project = myDebugProcess.getProject()

        if (DumbService.getInstance(project).isDumb()) return null

        return DebuggerUtils.findSourceFileForClass(project, GlobalSearchScope.allScope(project), className, sourceName, location.lineNumber() - 1)
    }

    override fun getAllClasses(sourcePosition: SourcePosition): List<ReferenceType> {
        if (sourcePosition.getFile() !is JetFile) {
            throw NoDataException()
        }
        val name = classNameForPosition(sourcePosition)
        val result = ArrayList<ReferenceType>()
        if (name != null) {
            result.addAll(myDebugProcess.getVirtualMachineProxy().classesByName(name))
        }
        return result
    }

    private fun classNameForPosition(sourcePosition: SourcePosition): String? {
        return runReadAction {
            val file = sourcePosition.getFile() as JetFile
            val isInLibrary = LibraryUtil.findLibraryEntry(file.getVirtualFile(), file.getProject()) != null
            val typeMapper = if (!isInLibrary) prepareTypeMapper(file) else createTypeMapperForLibraryFile(sourcePosition.getElementAt(), file)
            getClassNameForElement(sourcePosition.getElementAt(), typeMapper, file, isInLibrary)
        }
    }

    private fun prepareTypeMapper(file: JetFile): JetTypeMapper {
        val key = createKeyForTypeMapper(file)

        var value: CachedValue<JetTypeMapper>? = myTypeMappers.get(key)
        if (value == null) {
            value = CachedValuesManager.getManager(file.getProject()).createCachedValue<JetTypeMapper>(
                    {() ->
                        val typeMapper = createTypeMapper(file, key.second)
                        CachedValueProvider.Result<JetTypeMapper>(typeMapper, PsiModificationTracker.MODIFICATION_COUNT)
                    }, false)

            myTypeMappers.put(key, value)
        }

        return value!!.getValue()
    }

    override fun locationsOfLine(type: ReferenceType, position: SourcePosition): List<Location> {
        if (position.getFile() !is JetFile) {
            throw NoDataException()
        }
        try {
            val line = position.getLine() + 1
            val locations = if (myDebugProcess.getVirtualMachineProxy().versionHigher("1.4"))
                type.locationsOfLine(DebugProcess.JAVA_STRATUM, null, line)
            else
                type.locationsOfLine(line)
            if (locations == null || locations.isEmpty()) throw NoDataException()
            return locations
        }
        catch (e: AbsentInformationException) {
            throw NoDataException()
        }
    }

    override fun createPrepareRequest(classPrepareRequestor: ClassPrepareRequestor, sourcePosition: SourcePosition): ClassPrepareRequest? {
        if (sourcePosition.getFile() !is JetFile) {
            throw NoDataException()
        }
        val className = classNameForPosition(sourcePosition)
        if (className == null) {
            return null
        }
        return myDebugProcess.getRequestsManager().createClassPrepareRequest(classPrepareRequestor, className.replace('/', '.'))
    }

    TestOnly
    public fun addTypeMapper(file: JetFile, typeMapper: JetTypeMapper) {
        val value = CachedValuesManager.getManager(file.getProject()).createCachedValue<JetTypeMapper>(
                { () -> CachedValueProvider.Result<JetTypeMapper>(typeMapper, PsiModificationTracker.MODIFICATION_COUNT) }, false)
        val key = createKeyForTypeMapper(file)
        myTypeMappers.put(key, value)
    }

    class object {
        public fun createTypeMapper(file: JetFile, moduleInfo: IdeaModuleInfo): JetTypeMapper {
            val project = file.getProject()
            val packageFacadeScope = moduleInfo.contentScope()
            val packageFiles = PackageIndexUtil.findFilesWithExactPackage(file.getPackageFqName(), packageFacadeScope, project)

            val analysisResult = KotlinCacheService.getInstance(project).getAnalysisResults(packageFiles)
            analysisResult.throwIfError()

            val state = GenerationState(project, ClassBuilderFactories.THROW_EXCEPTION, analysisResult.moduleDescriptor,
                                        analysisResult.bindingContext, packageFiles.toList())
            state.beforeCompile()
            return state.getTypeMapper()
        }

        public fun getClassNameForElement(notPositionedElement: PsiElement?, typeMapper: JetTypeMapper, file: JetFile, isInLibrary: Boolean): String? {
            val element = getElementToCalculateClassName(notPositionedElement)
            when {
                element is JetClassOrObject -> return getJvmInternalNameForImpl(typeMapper, element)
                element is JetFunctionLiteral -> {
                    if (isInlinedLambda(element, typeMapper.getBindingContext())) {
                        return getClassNameForElement(element.getParent(), typeMapper, file, isInLibrary)
                    }
                    else {
                        val asmType = asmTypeForAnonymousClass(typeMapper.getBindingContext(), element)
                        return asmType.getInternalName()
                    }
                }
                element is JetClassInitializer -> {
                    val parent = getElementToCalculateClassName(element.getParent())
                    // Class-object initializer
                    if (parent is JetObjectDeclaration && parent.isClassObject()) {
                        return getClassNameForElement(parent.getParent(), typeMapper, file, isInLibrary)
                    }
                    return getClassNameForElement(element, typeMapper, file, isInLibrary)
                }
                element is JetProperty && (!element.isTopLevel() || !isInLibrary) -> {
                    if (isInPropertyAccessor(notPositionedElement)) {
                        val classOrObject = PsiTreeUtil.getParentOfType(element, javaClass<JetClassOrObject>())
                        if (classOrObject != null) {
                            return getJvmInternalNameForImpl(typeMapper, classOrObject)
                        }
                    }

                    val descriptor = typeMapper.getBindingContext().get(BindingContext.DECLARATION_TO_DESCRIPTOR, element)
                    if (descriptor !is PropertyDescriptor) {
                        return getClassNameForElement(element.getParent(), typeMapper, file, isInLibrary)
                    }

                    return getJvmInternalNameForPropertyOwner(typeMapper, descriptor)
                }
                element is JetNamedFunction -> {
                    val parent = getElementToCalculateClassName(element)
                    if (parent is JetClassOrObject) {
                        return getJvmInternalNameForImpl(typeMapper, parent)
                    }
                    else if (parent != null) {
                        val asmType = asmTypeForAnonymousClass(typeMapper.getBindingContext(), element)
                        return asmType.getInternalName()
                    }
                }
            }

            if (isInLibrary) {
                val elementAtForLibraryFile = getElementToCreateTypeMapperForLibraryFile(notPositionedElement)
                assert(elementAtForLibraryFile != null) {
                    "Couldn't find element at breakpoint for library file " + file.getName() +
                         (if (notPositionedElement == null) "" else ", notPositionedElement = " + JetPsiUtil.getElementTextWithContext(notPositionedElement))
                }
                return findPackagePartInternalNameForLibraryFile(elementAtForLibraryFile!!)
            }

            return PackagePartClassUtils.getPackagePartInternalName(file)
        }

        private fun getElementToCalculateClassName(notPositionedElement: PsiElement?) =
            PsiTreeUtil.getParentOfType(notPositionedElement,
                                        javaClass<JetClassOrObject>(),
                                        javaClass<JetFunctionLiteral>(),
                                        javaClass<JetNamedFunction>(),
                                        javaClass<JetProperty>(),
                                        javaClass<JetClassInitializer>())

        public fun getJvmInternalNameForPropertyOwner(typeMapper: JetTypeMapper, descriptor: PropertyDescriptor): String {
            return typeMapper.mapOwner(
                    if (AsmUtil.isPropertyWithBackingFieldInOuterClass(descriptor)) descriptor.getContainingDeclaration() else descriptor,
                    true).getInternalName()
        }

        private fun isInPropertyAccessor(element: PsiElement?) =
                element is JetPropertyAccessor ||
                PsiTreeUtil.getParentOfType(element, javaClass<JetProperty>(), javaClass<JetPropertyAccessor>()) is JetPropertyAccessor

        private fun getElementToCreateTypeMapperForLibraryFile(element: PsiElement?) =
                if (element is JetElement) element else PsiTreeUtil.getParentOfType(element, javaClass<JetElement>())

        private fun getJvmInternalNameForImpl(typeMapper: JetTypeMapper, jetClass: JetClassOrObject): String? {
            val classDescriptor = typeMapper.getBindingContext().get<PsiElement, ClassDescriptor>(BindingContext.CLASS, jetClass)
            if (classDescriptor == null) {
                return null
            }

            if (jetClass is JetClass && jetClass.isTrait()) {
                return typeMapper.mapTraitImpl(classDescriptor).getInternalName()
            }

            return typeMapper.mapClass(classDescriptor).getInternalName()
        }

        private fun createTypeMapperForLibraryFile(notPositionedElement: PsiElement?, file: JetFile): JetTypeMapper {
            val element = getElementToCreateTypeMapperForLibraryFile(notPositionedElement)
            val analysisResult = element!!.analyzeAndGetResult()

            val state = GenerationState(file.getProject(), ClassBuilderFactories.THROW_EXCEPTION,
                                        analysisResult.moduleDescriptor, analysisResult.bindingContext, listOf(file))
            state.beforeCompile()
            return state.getTypeMapper()
        }

        public fun isInlinedLambda(functionLiteral: JetFunctionLiteral, context: BindingContext): Boolean {
            val functionLiteralExpression = functionLiteral.getParent()
            if (functionLiteralExpression == null) return false

            var parent = functionLiteralExpression.getParent()

            var valueArgument: PsiElement = functionLiteralExpression
            while (parent is JetParenthesizedExpression || parent is JetBinaryExpressionWithTypeRHS || parent is JetLabeledExpression) {
                valueArgument = parent
                parent = parent.getParent()
            }

            while (parent is ValueArgument || parent is JetValueArgumentList) {
                parent = parent.getParent()
            }

            if (parent !is JetElement) return false

            val call = (parent as JetElement).getResolvedCall(context)
            if (call == null) return false

            val inlineType = InlineUtil.getInlineType(call.getResultingDescriptor())
            if (!inlineType.isInline()) return false

            for ((valueParameterDescriptor, resolvedValueArgument) in call.getValueArguments()) {
                for (next in resolvedValueArgument.getArguments()) {
                    val expression = next.getArgumentExpression()
                    if (valueArgument == expression) {
                        return InlineAnalyzerExtension.checkInlinableParameter(valueParameterDescriptor, expression, call.getResultingDescriptor(), null)
                    }
                }
            }
            return false
        }

        private fun createKeyForTypeMapper(file: JetFile) = Pair(file.getPackageFqName(), file.getModuleInfo())
    }
}
